### PR TITLE
SRP Batcher compatibility diagnostic

### PR DESCRIPTION
**Pull request by IgorMaliukh**
### Description

Implemented a new diagnostic to report if a shader is not compatible with the SRP batcher.

### Changes made

Added Shader SRP Batcher analyzer.

### Notes

Note that this should only be reported if the project is actually using a Scriptable Render Pipeline.

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
- [x] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
- [x] Docs for new/changed API's.
    - Code is annotated using Xmldoc syntax.

### CODE REVIEW

File: Editor/Modules/ShaderAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    {
+
+        internal const string PAS0000 = nameof(PAS0000);
+
+        internal static readonly Descriptor k_SrpBatcherDescriptor = new Descriptor(
+            PAS0000,
+            "Shader: Not compatible with SRP batcher",
+            Area.GPU,
+            "The shader is not compatible with SRP batcher.",
+            "Consider fixing the shader, if the SRP batcher compatibility was not intentionally removed."
```
Comment (mtrive) [inline Editor/Modules/ShaderAnalyzer.cs:20]: I would rephrase as "Consider adding SRP batcher compatibility to the shader. This will reduce the CPU time Unity requires to prepare and dispatch draw calls for materials that use the same shader variant."

File: Editor/Modules/ShaderAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        )
+        {
+            messageFormat = "Shader '{0}' is not compatible with SRP batcher.",
```
Comment (mtrive) [inline Editor/Modules/ShaderAnalyzer.cs:23]: messages don't require fullstop at the end of the sentence

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ## [0.9.3-preview.3] - 2023-02-28
 * Fixed lines and bars drawing
 * Fixed missing "Read/Write" diagnostic recommendations
+* Added Shader SRP Batcher analyzer.
```
[the remark this answers is not in the document]
Comment (IgorMaliukh) [reply CHANGELOG.md:10]: Done

File: Editor/Modules/ShaderAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+{
+    class ShaderAnalyzer : IShaderModuleAnalyzer
+    {
+
+        internal const string PAS0000 = nameof(PAS0000);
+
+        internal static readonly Descriptor k_SrpBatcherDescriptor = new Descriptor(
+            PAS0000,
+            "Shader: Not compatible with SRP batcher",
+            Area.GPU,
```
[the remark this answers is not in the document]
Comment (IgorMaliukh) [reply Editor/Modules/ShaderAnalyzer.cs:18]: Yes, you right

File: Editor/Modules/ShaderAnalyzer.cs (hunk 1 above)
Comment (IgorMaliukh) [reply Editor/Modules/ShaderAnalyzer.cs:20]: Done

File: Editor/Modules/ShaderAnalyzer.cs (hunk 2 above)
Comment (IgorMaliukh) [reply Editor/Modules/ShaderAnalyzer.cs:23]: Done

File: Tests/Editor/ShadersAnalysisTests.cs [hunk 5]
```diff
@@ -662,5 +705,23 @@ public void ShadersAnalysis_EditorDefaultResourcesShader_IsNotReported()
 
             Assert.Zero(issues.Length);
         }
+
+        [Test]
+#if !UNITY_2019_3_OR_NEWER
+        [Ignore("This requires the new Shader API")]
+#endif
+        public void ShadersAnalysis_SRPNonCompatibleShader_IsReported()
```
[the remark this answers is not in the document]
Comment (IgorMaliukh) [reply Tests/Editor/ShadersAnalysisTests.cs:713]: Yes, will add it

[the remark this answers is not in the document]
Comment (IgorMaliukh) [reply Tests/Editor/ShadersAnalysisTests.cs:757]: Done

**Review by mtrive — APPROVED** [review]
Great, thank you!
